### PR TITLE
PR for design discussion for  feat: SortKey struct now includes col IDs

### PR DESCRIPTION
**Pull request by NGA-TRAN**
This is the work/design proposal for https://github.com/influxdata/influxdb_iox/issues/8445
This is a task of https://github.com/influxdata/influxdb_iox/issues/6401

I do not plan to merge this PR butI need to make it compiled to see all the places I need to modify and start a discussion about the design.

### Main change:
Add col_ids in `SortKey` data structure which leads to changes in many places in ingester and querier

### Purpose of this PR:
- **Not** for commit/merge into main 
- For discovering all places that need to be changed when we modify `SortKey` data structure (aka the code compiles)

### What need to discuss?
- With everyone: 
      - to see if the new data structure `columns: Arc<IndexMap<Arc<str>, (i64, SortOptions)>>,`  in `SortKey` is good. This is the one Marco and I suggested.
      - to see if we should throw a panic when we cannot find column for a sort key string. I have opened https://github.com/influxdata/idpe/issues/18000 as a pre-work to avoid this panic
- With @crepererum: to see if adding `column_name_id_map: HashMap<Arc<str>, ColumnId>` into `QuerierParquetChunkMeta` is a feasible solution.
- With @domodwyer and @carols10cents: to see whether fetching cols in those 2 places in Ingester are the right places
- With @alamb: to see if fetching cols for import_export is at the right place
- Note: no discussion needed in compactor because there is no significant changes in Compactor as it does not update sort key.

### What have been done in this PR:
- The code compiles! Everything use SortKey works
- The partition's catalog field `sort_key_ids` is updated with appropriate values during catalog partition insert and update (no tests yet). I will split this part with new tests in a separate PR but I want to add it here for our discussion.

### What have NOT yet been done in this PR:
- Make all tests work (a lot of tests do not compile) but I do not want to touch them until we agree on sub-changes of data structures
- Tests for new functions. Similarly, I do not want to test them until we agree that they are good to go

 

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: import_export/src/file/import.rs [hunk 1]
```diff
@@ -522,13 +522,18 @@ impl RemoteImporter {
             return Err(Error::MismatchedSortKey { exported, existing });
         }
 
+        // fecth table columns
+        let columns = ColumnsByName::new(repos.columns().list_by_table_id(table.id).await?);
+        let new_sort_key_col_ids = columns.ids_for_names(&new_sort_key);
```
Comment (NGA-TRAN) [inline import_export/src/file/import.rs:527]: @alamb : I think this is the right place to fetch for this column ids but let me know what you think

File: querier/src/parquet/mod.rs [hunk 2]
```diff
@@ -26,6 +26,13 @@ pub struct QuerierParquetChunkMeta {
 
     /// Partition identifier.
     partition_id: TransitionPartitionId,
+
+    // TODO: I am not that happy storing table column name-id map in chunk
+    // metadata. It is not really metadata of the chunk, but rather of the table
+    // Ask Marco a better place for this
+    //
+    /// A map of column name to column ID for all columns of this chunk's table
+    column_name_id_map: HashMap<Arc<str>, ColumnId>,
```
Comment (NGA-TRAN) [inline querier/src/parquet/mod.rs:35]: @crepererum : The most important question for you here. Where should we store this to be able to get table col_ids for the changes of this PR  in `iox_query` and `querier`?

File: compactor/src/components/df_planner/query_chunk.rs [hunk 3]
```diff
@@ -108,6 +108,14 @@ impl QueryChunk for QueryableParquetChunk {
...
         self
     }
+
+    fn column_name_id_map(&self) -> HashMap<Arc<str>, ColumnId> {
+        // Return empty map as we don't need it for QueryableParquetChunk
+        // Reason: The column_name_id map is used to get ids of new column name.
+        // Since compactor's chunk do not add any new column and the chunk can
+        // even have fewer columns, we don't need the map.
+        HashMap::new()
+    }
```
Comment (NGA-TRAN) [inline compactor/src/components/df_planner/query_chunk.rs:118]: Note: if Marco points out we won't need to store table cols in Querier's chunk, this function won't be needed

File: data_types/src/columns.rs [hunk 4]
```diff
@@ -90,6 +91,14 @@ impl ColumnsByName {
         self.0.values().map(|c| c.id)
     }
 
+    /// Return an iterator of the set of column IDs of given names
```
Comment (NGA-TRAN) [inline data_types/src/columns.rs:94]:
```suggestion
    /// Return column IDs of given names
```

File: ingester/src/persist/worker.rs [hunk 5]
```diff
@@ -174,15 +175,26 @@ async fn compact_and_upload<O>(
 where
     O: Send + Sync,
 {
-    let compacted = compact(ctx, worker_state).await;
-    let (sort_key_update, parquet_table_data) = upload(ctx, worker_state, compacted).await;
+    // load sort key
+    let sort_key = ctx.sort_key().get().await;
+    // fetch column map
+    // THIS MUST BE DONE AFTER THE SORT KEY IS LOADED
+    let columns = fetch_column_map(ctx, worker_state, sort_key.clone()).await?;
```
Comment (NGA-TRAN) [inline ingester/src/persist/worker.rs:182]: @domodwyer and @carols10cents : Here is the first main change in ingester

File: ingester/src/persist/handle.rs [hunk 6]
```diff
@@ -432,8 +433,22 @@ impl PersistQueue for PersistHandle {
...
                 let data_primary_key = schema.primary_key();
-                if let Some(new_sort_key) = adjust_sort_key_columns(&v, &data_primary_key).1 {
+                // Verify if there are new column in PK
+                let new_columns = data_primary_key
+                    .iter()
+                    .filter(|c| !v.contains(c))
+                    .collect::<Vec<_>>();
+                if !new_columns.is_empty() {
                     // This persist operation will require a sort key update.
+                    // Nga TODO: Ask Dom whether to fetch table column map here or earlier at the same time fetching the sort key
```
Comment (NGA-TRAN) [inline ingester/src/persist/handle.rs:443]: @domodwyer and @carols10cents : Here is the second main change in Ingester that I am not sure whether it is the right place. Also, how to fetch it? Directly or thru something like `SortKeyState`?

File: schema/src/sort.rs [hunk 7]
```diff
@@ -7,15 +7,27 @@ use arrow::{
...
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
 #[derive(Debug, Default)]
 pub struct SortKeyBuilder {
-    columns: IndexMap<Arc<str>, SortOptions>,
+    /// The columns in the sort key, each is a mapping
+    /// from column name to its columnId and sort options
+    columns: IndexMap<Arc<str>, (i64, SortOptions)>,
```
Comment (NGA-TRAN) [inline schema/src/sort.rs:30]: This is the data structure change that leads to all the changes in this PR

File: iox_query/src/lib.rs [hunk 8]
```diff
@@ -90,6 +90,9 @@ pub trait QueryChunk: Debug + Send + Sync + 'static {
 
     /// Return backend as [`Any`] which can be used to downcast to a specific implementation.
     fn as_any(&self) -> &dyn Any;
+
+    /// Return column name-id map for table of this chunk
```
Comment (crepererum) [inline iox_query/src/lib.rs:94]: why does this leak all the way into the query interface? The query can happily continue to use strings. Having IDs in the catalog is a PURE storage optimization.

File: schema/src/sort.rs (hunk 7 above)
Comment (crepererum) [reply schema/src/sort.rs:30]: I don't see why this is required. The mapping between IDs and strings is an optimization that mostly concerns reading and writing to the catalog. While I agree that we MAY want to lift that to other components at one point, doing it all now is too invasive. It's too complicated to review, merge, and test and adds unnecessary complexity.

File: import_export/src/file/import.rs (hunk 1 above)
Comment (alamb) [reply import_export/src/file/import.rs:527]: I agree it looks right to me

Ideally we would likely export the column ids too into .json too as part of the export, but we can do that as some follow on PR

**Review by alamb — COMMENTED** [review]
It is neat to see this project progressing

File: data_types/src/partition.rs [hunk 9]
```diff
@@ -385,6 +388,33 @@ pub struct Partition {
...
+    /// first persist when the sort key is empty, as on subsequent
+    /// persist operations when new tags occur in newly inserted data.
+    ///
+    /// Updating inserts new column into the existing order. The order
+    /// of the existing columns relative to each other is NOT changed.
+    ///
+    /// For example, updating `A,B,C` to either `A,D,B,C` or `A,B,C,D`
+    /// is legal. However, updating to `A,C,D,B` is not because the
+    /// relative order of B and C have been reversed.
+    pub sort_key_ids: Vec<ColumnId>,
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply data_types/src/partition.rs:416]: The plan is to remove `sort_key` after all `sort_key_ids` have values. If you see the tasks of the [feature ticke](https://github.com/influxdata/influxdb_iox/issues/6401), the last step is to drop `sort_key`

**Pull request closed without merging**